### DISCUSSION TITLE
Allow `unsafe` modifier in namespace declaration

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#5343] [Language Ideas]
`unsafe` blocks are often used in entire types, and most of the time, across the entire file. Allowing the modifier to be used in namespace declarations would mitigate that problem, and reduce the number of modifiers used in the type declaration.

Assuming that namespace declarations are implicitly partial, placing this modifier should work the same way it behaves in partial type declarations; which is block-wide, and not type-wide. This means that the `unsafe` modifier would only apply to the current file for file-scoped namespace declarations, or the current block for standard namespace declarations.

Example:
```csharp
// -- File0.cs
unsafe namespace N;

public class C
{
    public void M(void* ptr) { } // valid
}

// -- File1.cs
namespace N
{
    public class D
    {
        public void M(void* ptr) { } // error; requires unsafe modifier
    }
}
unsafe namespace N
{
    public class E
    {
        public void M(void* ptr) { } // valid
    }
}
```

This proposal obviously opens up possibilities for potentially permitting other modifiers in a similar fashion that would not alter behavior of namespace declarations.

## Alternatives

At second thought after maturing this proposal in my mind for months, I believe it would best suit as a standalone top-level statement before the first namespace/type declaration:
```csharp
using ...;

unsafe; // the whole file now has unsafe semantics enabled
// discerned from namespace semantics

namespace SomeNamespace;
// declarations using unsafe
```

The above proposal was not well-received, and there are great arguments against it, thus it may be considered discarded. The proposed alternative has to live somewhere, either by transforming this current discussion for it, or creating a brand new discussion to accommodate the changes.

**Comment by HaloFour** [comment]
Considering most non-trivial code is written one class per file this seems to do nothing more than move where the `unsafe` modifier should be specified.

**Reply by CyrusNajmabadi** [reply]
This would also likely be too confusing as `unsafe` is lexically scoped, but Namespaces are normally not. There is a little overlap here with partial types. However partial types are the exception, while namespaces always work this way.

**Reply by Rekkonnect** [reply]
I will have to agree that `unsafe` in front of `namespace` more or less has an ambiguous meaning, when it'll be made to declare the scope as unsafe, however partial types work exactly the same and provide the exact same ambiguity. The good part about this is that `unsafe` is almost exclusively used by already experienced users, so having to learn how this works will be simple, and they won't be in pain.

And if they ever ask namespace-scoped unsafety, `global unsafe namespace` is always (never) a possibility.

**Reply by Rekkonnect** [reply]
Another argument in favor of this feature is that users that use `unsafe` will be more experienced with how `unsafe` exactly works in partial declarations, causing less confusion every time they use both `unsafe` and `partial` in type declarations, since the behavior will correlate to that of an `unsafe namespace`.

**Reply by HaloFour** [reply]
It's not the type that is unsafe, it's the scope.  The lexical scoping thus makes sense with partial types.  I'm not saying it wouldn't with namespaces as well.

But I don't see where the value lies in adding the `unsafe` modifier at the `namespace` level.  At best it means a few less keywords if, and only if, you happen to pack a bunch of `unsafe` types into a single file under a single namespace.  I'd argue that this would be a bad practice and not one that should be encouraged by the language, and even then the savings is trivial at best.

**Comment by Xyncgas** [comment]
let's make this work guys

**Comment by huoyaoyuan** [comment]
`class` needs to be `unsafe` because it can directly contain pointer fields.
Methods need to be unsafe because it can contain pointer parameters.

**Reply by Rekkonnect** [reply]
An unsafe type could contain no pointer fields but enables all its members, methods, fields, properties and nested types to use pointer types and unsafe features (fixed, SkipLocalsInit).

I don't quite get what you're trying to convey, if I'm being honest.

**Comment by CyrusNajmabadi** [comment]
Given that one type per file is the 99.9 case (we've actually measured this), there seems to be extremely marginal gain here. In nearly all cases you just put `unsafe` on the type and you're good. 

For the rare case of multiple types, requiring the modifier be on the type seems fine.
